src/Orchard/Tasks/Locking/Services/DistributedLock.cs

```diff
@@ -3,44 +3,34 @@
 
 namespace Orchard.Tasks.Locking.Services {
 
-    /// <summary>
-    /// Represents a distributed lock. />
-    /// </summary>
-    public class DistributedLock : IDisposable {
-        public static DistributedLock ForMachine(IDistributedLockService service, string name, string machineName, string lockId) {
-            return new DistributedLock {
-                _service = service,
-                Name = name,
-                MachineName = machineName,
-                Id = lockId
-            };
-        }
+    public class DistributedLock : IDistributedLock {
 
-        public static DistributedLock ForThread(IDistributedLockService service, string name, string machineName, int threadId, string lockId) {
-            return new DistributedLock {
-                _service = service,
-                Name = name,
-                MachineName = machineName,
-                ThreadId = threadId,
-                Id = lockId
-            };
-        }
+        private DistributedLockService _service;
+        private string _name;
+        private int _count;
 
-        private IDistributedLockService _service;
-        private int _isDisposed;
+        public string Name {
+            get {
+                return _name;
+            }
+        }
 
-        private DistributedLock() {
+        public DistributedLock(DistributedLockService service, string name) {
+            _service = service;
+            _name = name;
+            _count = 1;
         }
 
-        public string Id { get; private set; }
-        public string Name { get; private set; }
-        public string MachineName { get; private set; }
-        public int? ThreadId { get; private set; }
+        public void IncreaseReferenceCount() {
+            _count++;
+        }
 
-        // This will be called at least and at the latest by the IoC container when the request ends.
         public void Dispose() {
-            if(Interlocked.CompareExchange(ref _isDisposed, 1, 0) == 0)
-                _service.ReleaseLock(this);
+            _count--;
+            if (_count == 0) {
```

Comment:
Why? How could two thread use the same lock object? They are created per dependency, so per request, so two threads can't call dispose on the same lock.